File: Food Ordering Application API/Business_Layer/Services/RestaurantService.cs

using Data_Accesss_Layer.Models;
using Data_Accesss_Layer.Repositories;

namespace Business_Layer.Services
{
    public class RestaurantService : IRestaurantService
    {
        private readonly IRestaurantRepository _restaurantRepository;

        public RestaurantService(IRestaurantRepository restaurantRepository)
        {
            _restaurantRepository = restaurantRepository;
        }

        public Restaurant RegisterRestaurant(int ownerId, string name, string location, string cuisine)
        {
            var restaurant = new Restaurant
            {
                OwnerId = ownerId,
                Name = name,
                Location = location,
                Cuisine = cuisine
            };

            _restaurantRepository.AddRestaurant(restaurant);
            _restaurantRepository.SaveChanges();
            return restaurant;
        }

        public Menu AddMenuItem(int restaurantId, string name, decimal price)
        {
            var menuItem = new Menu
            {
                RestaurantId = restaurantId,
                Name = name,
                Price = price
            };

            _restaurantRepository.AddMenuItem(menuItem);
            _restaurantRepository.SaveChanges();
            return menuItem;
        }
        public Menu GetMenuItemById(int menuId)
        {
            return _restaurantRepository.GetMenuItemById(menuId);
        }
        public Menu EditMenuItem(int menuId, string name, decimal price)
        {
            var menuItem = _restaurantRepository.GetMenuItemById(menuId);
            if (menuItem == null)
                throw new KeyNotFoundException("Menu item not found.");

            menuItem.Name = name;
            menuItem.Price = price;

            _restaurantRepository.UpdateMenuItem(menuItem);
            _restaurantRepository.SaveChanges();
            return menuItem;
        }

        public void DeleteMenuItem(int menuId)
        {
            var menuItem = _restaurantRepository.GetMenuItemById(menuId);
            if (menuItem == null)
                throw new KeyNotFoundException("Menu item not found.");

            _restaurantRepository.DeleteMenuItem(menuItem);
            _restaurantRepository.SaveChanges();
        }

        public List<Review> ViewReviews(int restaurantId)
        {
            return _restaurantRepository.GetReviewsByRestaurantId(restaurantId);
        }

        public void RespondToReview(int reviewId, string response)
        {
            var review = _restaurantRepository.GetReviewById(reviewId);
            if (review == null)
                throw new KeyNotFoundException("Review not found.");

            
            review.Response = response;
            _restaurantRepository.SaveChanges();
        }
    }
}

File: Food_Ordering_Application_UI/src/app/services/restaurant.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Restaurant } from '../models/Restaurant';
import { Observable } from 'rxjs/internal/Observable';
import { Menu } from '../models/Menu';
import { Review } from '../models/Review';

@Injectable({
  providedIn: 'root'
})
export class RestaurantService {
  
  private apiUrl = 'https://localhost:7090/api/RestaurantOwner'; 
  private api='https://localhost:7090/api/Admin';
  constructor(private http: HttpClient) {}

  registerRestaurant(restaurantDto: {ownerId: number, name: string, location: string, cuisine: string}): Observable<Restaurant> {
    return this.http.post<Restaurant>(`${this.apiUrl}/registerRestaurant`, restaurantDto);
  }
  getAllRestaurants(): Observable<Restaurant[]> {
    return this.http.get<Restaurant[]>(`${this.api}/getAllRestaurants`);
  }
  // addMenuItem(restaurantId: number, name: string, price: number): Observable<Menu> {
  //   return this.http.post<Menu>(`${this.apiUrl}/addMenuItem`, { restaurantId, name, price });
  // }
  addMenuItem(menuItemDto: { restaurantId: number; name: string; price: number }): Observable<Menu> {
    return this.http.post<Menu>(`${this.apiUrl}/addMenuItem`, menuItemDto);
  }
  
  // editMenuItem(menuId: number, name: string, price: number): Observable<Menu> {
  //   return this.http.put<Menu>(`${this.apiUrl}/editMenuItem/${menuId}`, { name, price });
  // }


  // Edit an existing menu item
  editMenuItem(menuId: number, menuItemDto: { name: string; price: number }): Observable<Menu> {
    return this.http.put<Menu>(`${this.apiUrl}/editMenuItem/${menuId}`, menuItemDto);
  }
  deleteMenuItem(menuId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/deleteMenuItem/${menuId}`);
  }
 
  viewReviews(restaurantId: number): Observable<Review[]> {
    return this.http.get<Review[]>(`${this.apiUrl}/viewReviews/${restaurantId}`);
  }
  getMenuItemById(menuId: number): Observable<Menu> {
    return this.http.get<Menu>(`${this.apiUrl}/menuItem/${menuId}`);
  }
  // respondToReview(reviewId: number, response: string): Observable<void> {
  //   return this.http.post<void>(`${this.apiUrl}/respondToReview`, { reviewId, response });
  // }

  // Respond to a specific review
  respondToReview(reviewDto: { reviewId: number; response: string }): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/respondToReview`, reviewDto);
  }
}
